src/backend/distributed/utils/metadata_cache.c

```diff
@@ -194,13 +195,68 @@ IsDistributedTable(Oid relationId)
 		return false;
 	}
 
-	cacheEntry = LookupDistTableCacheEntry(relationId);
+	/*
+	 * We will not use cache if there is a version mismatch, because while
+	 * creating the cache, we assume our metadata tables are in particular
+	 * structure. However, in version mismatch, this is not guaranteed, thus
+	 * we will read from pg_dist_partition table and we will only assume
+	 * specific details which are most likely do not change in long term.
+	 */
+	installedVersion = InstalledExtensionVersion();
+	if (!MajorVersionsCompatible(installedVersion, CITUS_EXTENSIONVERSION))
```

Comment:
This doesn't look good to me, InstalledExtensionVersion() is far from cheap.  I think you're going to have to integrate this with the state kept in CitusHasBeenLoaded(), i.e. once determined that things are compatible, it needs to be cached.